### PR TITLE
Qutrit channel operation

### LINKED ISSUE #5649: Add a QutritChannel operation
### Feature details

Version 0.36 of PennyLane includes a [`"default.qutrit.mixed"`](https://docs.pennylane.ai/en/latest/code/api/pennylane.devices.default_qutrit_mixed.DefaultQutritMixed.html) device that is capable of simulating [qutrit systems](https://pennylane.ai/qml/demos/tutorial_qutrits_bernstein_vazirani/) in the mixed-state setting. This unlocks the ability to simulate qutrit-based noise, and we would like to add more qutrit noise channels to complement our existing range of [qutrit operators](https://docs.pennylane.ai/en/stable/introduction/operations.html#qutrit-operators).

This objective of this issue is to provide a qutrit analogue of [`QubitChannel`](https://docs.pennylane.ai/en/stable/code/api/pennylane.QubitChannel.html), which would allow noise to be specified via a collection of $(3\times 3)$-dimensional [Kraus operators](https://en.wikipedia.org/wiki/Quantum_operation#Kraus_operators).

### Implementation

It should be possible to execute the following code:

```python
import pennylane as qml
import numpy as np

kraus = [  # See https://github.com/PennyLaneAI/pennylane/pull/5503
    np.array([[1, 0, 0],
              [0, 0.70710678, 0],
              [0, 0, 0.8660254]]),
    np.array([[0, 0.70710678, 0],
              [0, 0, 0],
              [0, 0, 0]]),
    np.array([[0, 0, 0.5],
              [0, 0, 0],
              [0, 0, 0]])
]

dev = qml.device("default.qutrit.mixed")

@qml.qnode(dev)
def f():
    qml.QutritChannel(kraus, 0)
    return qml.expval(qml.GellMann(wires=0, index=1))

f()
```

This can be done by following these instructions:

1. Implement a `qml.QutritChannel` operation in `pennylane/ops/channel.py`, which defines the qutrit equivalent of the `qml.QubitChannel` operation. This operation should:
    * Take a list containing three $3 \times 3$ Kraus matrices, along with `wires` and `id` (optional string representation of the operator).
    * Define the `compute_kraus_matrices`, which is a `staticmethod` that returns the Kraus matrices in a list, and `_flatten` functions (just like in `qml.QubitChannel`). 

2. Add unit and integration tests:
    * Unit tests must verify the correctness of the operation, its attributes, and its methods.
    * Integration tests must show that the output of a `QNode` that applies the `qml.QutritChannel` operation is correct.
    * Integration tests must also verify that `QNode`s that apply the `qml.QutritChannel` operation can be differentiated using the `parameter-shift`, `finite-diff`, and `backprop` differentiation methods and that the computed gradients/jacobians are correct.
    * Integration tests must be implemented for both the `default.qutrit` and `default.qutrit.mixed` devices.

A successful PR must include:

- The requested feature
- Comprehensive tests
- Relevant updates to the documentation
- An entry in the [release notes](https://github.com/PennyLaneAI/pennylane/blob/master/doc/releases/changelog-dev.md)
- Passing all CI tests

Check out our [development guide](https://docs.pennylane.ai/en/stable/development/guide.html) for more information, and don't hesitate to reach out to us if you need assistance.

### How important would you say this feature is?

1: Not important. Would be nice to have.

### Additional information

Existing work is underway to support specific qutrit channels in PRs https://github.com/PennyLaneAI/pennylane/pull/5502 and https://github.com/PennyLaneAI/pennylane/pull/5503.

### PR DESCRIPTION
**Pull request by Tarun-Kumar07**
**Context:**
The goal of this PR is to create a qutrit counterpart to the QubitChannel, enabling noise to be defined using a set of 
3×3 Kraus operators.

**Description of the Change:**
- Implement `QutritChannel` similar to `QubitChannel`

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**
Fixes #5649

### CODE REVIEW

File: pennylane/ops/qutrit/channel.py [hunk 1]
```diff
@@ -313,3 +313,74 @@ def compute_kraus_matrices(gamma_1, gamma_2):  # pylint:disable=arguments-differ
...
+
+        # check that the channel represents a trace-preserving map
+        if not any(math.is_abstract(K) for K in K_list):
+            K_arr = math.array(K_list)
+            Kraus_sum = math.einsum("ajk,ajl->kl", K_arr.conj(), K_arr)
+            if not math.allclose(Kraus_sum, math.eye(K_list[0].shape[0])):
+                raise ValueError("Only trace preserving channels can be applied.")
+
+    def _flatten(self):
+        return (self.data,), (self.wires, ())
```
[the remark this answers is not in the document]
Comment (mudit2812) [reply pennylane/ops/qutrit/channel.py:394]: You can test it similar to [this test](https://github.com/PennyLaneAI/pennylane/blob/7384db3693cef7d04d2414e021714791a11b0d42/tests/test_operation.py#L326).

File: doc/releases/changelog-dev.md [hunk 2]
```diff
@@ -151,6 +151,8 @@
   
 * ``qml.QutritDepolarizingChannel`` has been added, allowing for depolarizing noise to be simulated on the `default.qutrit.mixed` device.
   [(#5502)](https://github.com/PennyLaneAI/pennylane/pull/5502)
+ 
+* `qml.QutritChannel` has been added, enabling the specification of noise using a collection of (3x3) Kraus matrices on the `default.qutrit.mixed` device. [(#5793)](https://github.com/PennyLaneAI/pennylane/issues/5649)
```
Comment (mudit2812) [inline doc/releases/changelog-dev.md:155]:
```suggestion
* `qml.QutritChannel` has been added, enabling the specification of noise using a collection of (3x3) Kraus matrices on the `default.qutrit.mixed` device.
  [(#5793)](https://github.com/PennyLaneAI/pennylane/issues/5649)
```

File: pennylane/ops/qutrit/channel.py [hunk 3]
```diff
@@ -340,3 +340,74 @@ def compute_kraus_matrices(gamma_1, gamma_2, gamma_3):  # pylint:disable=argumen
             math.cast_like(math.array([[0, 0, 0], [0, 0, 1], [0, 0, 0]]), gamma_3), gamma_3
         )
         return [K0, K1, K2, K3]
+
+
+class QutritChannel(Channel):
+    r"""
+    Apply an arbitrary fixed quantum channel.
```
Comment (mudit2812) [inline pennylane/ops/qutrit/channel.py:347]:
```suggestion
    Apply an arbitrary fixed qutrit channel.
```

File: pennylane/ops/qutrit/channel.py [hunk 4]
```diff
@@ -340,3 +340,74 @@ def compute_kraus_matrices(gamma_1, gamma_2, gamma_3):  # pylint:disable=argumen
...
+            raise ValueError(
+                "Only channels with the same input and output Hilbert space dimensions can be applied."
+            )
+
+        # check all Kraus matrices have the same shape
+        if any(K.shape != K_list[0].shape for K in K_list):
+            raise ValueError("All Kraus matrices must have the same shape.")
+
+        # check the dimension of all Kraus matrices are valid
+        if any(K.shape[0] != QUDIT_DIM for K in K_list):
```
Comment (mudit2812) [inline pennylane/ops/qutrit/channel.py:381]: This condition will raise an error if there are multiple wires since the expected shape is `(3 ** num_wires, 3 ** num_wires)`.

```suggestion
        if any(K.shape[0] != QUDIT_DIM ** (len(self.wires)) for K in K_list):
```

**Review by mudit2812 — COMMENTED** [review]
Thanks for opening the PR @Tarun-Kumar07 ! I've just left a couple of preliminary comments. I noticed that there aren't any tests for `QutritChannel`s with multiple wires. Could you include those as well? Additionally, I noticed that the tests call `channel.QutritChannel`, but this should be `qml.QutritChannel`.

Feel free to re-request review whenever the comments are resolved 😄

**Comment by Tarun-Kumar07** [comment]
Hi @mudit2812,

Thank you for the review. I have modified the unit test `test_kraus_matrices_are_dimensions` to ensure it runs on multiple wires. Additionally, I have adjusted the error message for clarity. Could you please review it again?

Thanks!

File: doc/releases/changelog-dev.md [hunk 5]
```diff
@@ -179,6 +179,9 @@
 
 * ``qml.QutritDepolarizingChannel`` has been added, allowing for depolarizing noise to be simulated on the `default.qutrit.mixed` device.
   [(#5502)](https://github.com/PennyLaneAI/pennylane/pull/5502)
+ 
+* `qml.QutritChannel` has been added, enabling the specification of noise using a collection of (3x3) Kraus matrices on the `default.qutrit.mixed` device.
+  [(#5793)](https://github.com/PennyLaneAI/pennylane/issues/5649)
```
Comment (trbromley) [inline doc/releases/changelog-dev.md:184]:
```suggestion
  [(#5793)](https://github.com/PennyLaneAI/pennylane/issues/5793)
```

File: pennylane/ops/qutrit/channel.py [hunk 6]
```diff
@@ -378,10 +378,9 @@ def __init__(self, K_list, wires=None, id=None):
             raise ValueError("All Kraus matrices must have the same shape.")
 
         # check the dimension of all Kraus matrices are valid
-        if any(K.shape[0] != QUDIT_DIM for K in K_list):
-            raise ValueError(
-                f"Dimension of all Kraus matrices must be ({QUDIT_DIM}**num_wires, {QUDIT_DIM}**num_wires)."
-            )
+        kraus_dim = QUDIT_DIM ** len(self.wires)
+        if any(K.shape[0] != kraus_dim for K in K_list):
+            raise ValueError(f"Dimension of all Kraus matrices must be ({kraus_dim},{kraus_dim}).")
```
Comment (mudit2812) [inline pennylane/ops/qutrit/channel.py:383]:
```suggestion
            raise ValueError(f"Shape of all Kraus matrices must be ({kraus_dim},{kraus_dim}).")
```

**Review by mudit2812 — COMMENTED** [review]
Looks great! I'll approve once CI is green. Nice work @Tarun-Kumar07

**Comment by Tarun-Kumar07** [comment]
Hi @mudit2812 and @trbromley , I am not sure why the test `upload-to-codecov` failed. Could you please help.
I have addressed rest of the code review comments, please let me know if something needs to improved.

Thanks !!

**Review by mudit2812 — CHANGES_REQUESTED** [review]
One final change I forgot to request is to add `QutritChannel` to the operations page in the pennylane docs. [This](https://docs.pennylane.ai/en/latest/introduction/operations.html#qutrit-noisy-channels) is the section to add the operator, and [this](https://github.com/PennyLaneAI/pennylane/blob/38b3e74e26ef7f4f758e1cb34d4db0a1fc679780/doc/introduction/operations.rst?plain=1#L510) is the file that needs to be updated to add the operation.

**Review by mudit2812 — APPROVED** [review]
Nice work @Tarun-Kumar07 ! I'll pass this on to another team member now for a second review.

File: tests/ops/qutrit/test_qutrit_channel_ops.py [hunk 7]
```diff
@@ -288,3 +288,98 @@ def test_kraus_jac_jax(self):
...
+    def test_flatten(self):
+        """Test flatten method returns kraus matrices and wires"""
+        kraus = [
+            np.array([[1, 0, 0], [0, 0.70710678, 0], [0, 0, 0.8660254]]),
+            np.array([[0, 0.70710678, 0], [0, 0, 0], [0, 0, 0]]),
+            np.array([[0, 0, 0.5], [0, 0, 0], [0, 0, 0]]),
+        ]
+
+        qutrit_channel = qml.QutritChannel(kraus, 1, id="test")
+        data, metadata = qutrit_channel._flatten()  # pylint: disable=protected-access
```
Comment (albi3ro) [inline tests/ops/qutrit/test_qutrit_channel_ops.py:382]:
```suggestion
        data, metadata = qutrit_channel._flatten()  # pylint: disable=protected-access
        new_op = qml.QutritChannel._unflatten(data, metadata)
        assert qml.equal(qutrit_channel, new_op)
```

We should probably make sure the flatten method is compatible with the analogous `unflatten` method.

File: pennylane/ops/qutrit/channel.py [hunk 8]
```diff
@@ -457,3 +457,73 @@ def compute_kraus_matrices(p_01, p_02, p_12):  # pylint:disable=arguments-differ
...
+            raise ValueError(f"Shape of all Kraus matrices must be ({kraus_dim},{kraus_dim}).")
+
+        # check that the channel represents a trace-preserving map
+        if not any(math.is_abstract(K) for K in K_list):
+            K_arr = math.array(K_list)
+            Kraus_sum = math.einsum("ajk,ajl->kl", K_arr.conj(), K_arr)
+            if not math.allclose(Kraus_sum, math.eye(K_list[0].shape[0])):
+                raise ValueError("Only trace preserving channels can be applied.")
+
+    def _flatten(self):
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/ops/qutrit/channel.py:509]: For context, `Operator._flatten` and `Operator._unflatten` are used to improve out integration with jax, see [jax pytrees docs](https://jax.readthedocs.io/en/latest/pytrees.html).  Now jitted functions can can accept and return operators, and we can have improved parameter handling.

The contraints for these two methods are:

1) `type(op)._unflatten(*op._flatten())` should return the original operation.
2) Any traceable/ trainable data or any other pytrees should be placed in the `data` output
3) the `metadata` output should be hashable and immutable.

An example of this in action is:
```
op = qml.adjoint(2*qml.X(0)) @ qml.Rot(1.2, 2.3, 3.4, 0)
data, structure = qml.pytrees.flatten(op)
qml.pytrees.unflatten([3.0, -1.2, -2.3, -3.4], structure)
```

[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/ops/qutrit/channel.py:509]: Ignore me.  

This is due to the difference between `QutritChannel((matrix1, matrix2), wires)` instead of `QutritChannel(matrix1, matrix2, wires)`.  Makes sense now.

**Review by albi3ro — APPROVED** [review]
🎉

**Comment by Tarun-Kumar07** [comment]
Hi @albi3ro and @mudit2812 ,

Can this PR be merged as there are two approvals.

Thanks !!

**Comment by Tarun-Kumar07** [comment]
Hey @mudit2812 , I am not sure why `upload-to-codecov` is failing :).

**Comment by mudit2812** [comment]
> Hey @mudit2812 , I am not sure why `upload-to-codecov` is failing :).

I've retriggered the upload to codecov job. Sometimes it happens 😅

**Comment by mudit2812** [comment]
Not sure what's going on but it's failing repeatedly. I'll bring this up with the team.

**Comment by mudit2812** [comment]
@Tarun-Kumar07 for now it might be worth just running CI all over again by pushing an empty commit. You can do so using `git commit --allow-empty -m "Trigger CI"`
